### PR TITLE
Updated alibaba instance types

### PR DESCRIPTION
**Pull request by kgroschoff**
**What this PR does / why we need it**:
Updated Alibaba instance types, as the old method seems not to filter response correctly by region. Therefor a lot of invalid instance types were in the list. 

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: api/pkg/handler/v1/provider/alibaba.go [hunk 1]
```diff
@@ -132,44 +132,56 @@ func AlibabaInstanceTypesWithClusterCredentialsEndpoint(projectProvider provider
...
 			return nil, err
 		}
 
-		return listAlibabaInstanceTypes(ctx, accessKeyID, accessKeySecret, req.Region)
+		return listAlibabaInstanceTypes(ctx, seedsGetter, accessKeyID, accessKeySecret, req.Region)
 	}
 }
 
-func listAlibabaInstanceTypes(ctx context.Context, accessKeyID string, accessKeySecret string, region string) (apiv1.AlibabaInstanceTypeList, error) {
+func listAlibabaInstanceTypes(ctx context.Context, seedsGetter provider.SeedsGetter, accessKeyID string, accessKeySecret string, region string) (apiv1.AlibabaInstanceTypeList, error) {
```
Comment (moadqassem) [inline api/pkg/handler/v1/provider/alibaba.go]: In this function the `seedsGetter` is not used or?

File: api/pkg/handler/v1/provider/alibaba.go [hunk 2]
```diff
@@ -154,34 +154,59 @@ func AlibabaInstanceTypesWithClusterCredentialsEndpoint(projectProvider provider
...
+	}
+
+	// get all instance types and filter afterwards, to reduce calls
+	requestInstanceTypes := ecs.CreateDescribeInstanceTypesRequest()
+	requestInstanceTypes.Scheme = "https"
+
+	instTypes, err := client.DescribeInstanceTypes(requestInstanceTypes)
 	if err != nil {
-		return apiv1.AlibabaInstanceTypeList{}, errors.New(http.StatusInternalServerError, fmt.Sprintf("failed to list instance types: %v", err))
+		fmt.Print(apiv1.AlibabaInstanceTypeList{}, errors.New(http.StatusInternalServerError, fmt.Sprintf("failed to list instance types: %#v", err)))
```
Comment (zreigz) [inline api/pkg/handler/v1/provider/alibaba.go]: should be return?

File: api/pkg/handler/v1/provider/alibaba.go [hunk 3]
```diff
@@ -154,34 +154,59 @@ func AlibabaInstanceTypesWithClusterCredentialsEndpoint(projectProvider provider
...
+		}
+	}
+	return false
+}
+
 func listAlibabaInstanceTypes(ctx context.Context, seedsGetter provider.SeedsGetter, accessKeyID string, accessKeySecret string, region string) (apiv1.AlibabaInstanceTypeList, error) {
+	// Alibaba has way too many instance types that are not all available in each region
+	// recommendedInstanceFamilies are those families that are recommended in this document:
+	// https://www.alibabacloud.com/help/doc-detail/25378.htm?spm=a2c63.p38356.b99.47.7acf342enhNVmo
+	recommendedInstanceFamilies := []string{"ecs.g6", "ecs.g5", "ecs.g5se", "ecs.g5ne", "ecs.ic5", "ecs.c6", "ecs.c5", "ecs.r6", "ecs.r5", "ecs.d1ne", "ecs.i2", "ecs.i2g", "ecs.hfc6", "ecs.hfg6", "ecs.hfr6"}
```
Comment (zreigz) [inline api/pkg/handler/v1/provider/alibaba.go]: Please use hash set for this: "k8s.io/apimachinery/pkg/util/sets"
```
recommendedInstanceFamilies := sets.NewString("ecs.g6", "ecs.g5", "ecs.g5se", "ecs.g5ne", "ecs.ic5", "ecs.c6", "ecs.c5", "ecs.r6", "ecs.r5", "ecs.d1ne", "ecs.i2", "ecs.i2g", "ecs.hfc6", "ecs.hfg6", "ecs.hfr6")
```
Then you can remove `isInList` method. `recommendedInstanceFamilies.Has(instType.InstanceTypeFamily)`

File: api/pkg/handler/v1/provider/alibaba.go [hunk 4]
```diff
@@ -154,34 +154,59 @@ func AlibabaInstanceTypesWithClusterCredentialsEndpoint(projectProvider provider
...
-
-	instTypes, err := client.DescribeAccountAttributes(request)
+	// get all families that are available for the Region
+	requestFamilies := ecs.CreateDescribeInstanceTypeFamiliesRequest()
+	requestFamilies.Scheme = "https"
+	requestFamilies.RegionId = region
+
+	instTypeFamilies, err := client.DescribeInstanceTypeFamilies(requestFamilies)
+	if err != nil {
+		fmt.Print(apiv1.AlibabaInstanceTypeList{}, errors.New(http.StatusInternalServerError, fmt.Sprintf("failed to list instance type families: %#v", err)))
```
Comment (zreigz) [inline api/pkg/handler/v1/provider/alibaba.go]: return ?

**Review by zreigz — CHANGES_REQUESTED** [review]

File: api/pkg/handler/v1/provider/alibaba.go [hunk 5]
```diff
@@ -7,6 +7,7 @@ import (
 
 	"github.com/aliyun/alibaba-cloud-sdk-go/services/ecs"
 	"github.com/go-kit/kit/endpoint"
+	"k8s.io/apimachinery/pkg/util/sets"
```
Comment (moadqassem) [inline api/pkg/handler/v1/provider/alibaba.go]: can you please have better regrouping for the import paths? `"k8s.io/apimachinery/pkg/util/sets"` comes after `kubermatic` imports.

File: api/pkg/handler/v1/provider/alibaba.go [hunk 6]
```diff
@@ -132,40 +133,68 @@ func AlibabaInstanceTypesWithClusterCredentialsEndpoint(projectProvider provider
...
-	request := ecs.CreateDescribeInstanceTypesRequest()
-	request.Scheme = "https"
+	// get all families that are available for the Region
+	requestFamilies := ecs.CreateDescribeInstanceTypeFamiliesRequest()
+	requestFamilies.Scheme = "https"
+	requestFamilies.RegionId = region
+
+	instTypeFamilies, err := client.DescribeInstanceTypeFamilies(requestFamilies)
+	if err != nil {
+		return nil, errors.New(http.StatusInternalServerError, fmt.Sprintf("failed to list instance type families: %#v", err))
```
Comment (moadqassem) [inline api/pkg/handler/v1/provider/alibaba.go]: Why formating the error message with `#` ?

File: api/pkg/handler/v1/provider/alibaba.go [hunk 7]
```diff
@@ -132,40 +133,68 @@ func AlibabaInstanceTypesWithClusterCredentialsEndpoint(projectProvider provider
...
+
+	// get all instance types and filter afterwards, to reduce calls
+	requestInstanceTypes := ecs.CreateDescribeInstanceTypesRequest()
+	requestInstanceTypes.Scheme = "https"
 
-	instTypes, err := client.DescribeInstanceTypes(request)
+	instTypes, err := client.DescribeInstanceTypes(requestInstanceTypes)
 	if err != nil {
-		return apiv1.AlibabaInstanceTypeList{}, errors.New(http.StatusInternalServerError, fmt.Sprintf("failed to list instance types: %v", err))
+		return nil, errors.New(http.StatusInternalServerError, fmt.Sprintf("failed to list instance types: %#v", err))
```
Comment (moadqassem) [inline api/pkg/handler/v1/provider/alibaba.go]: Same goes here, Why formating the error message with `#` ?

**Comment by kgroschoff** [comment]
Thanks @zreigz & @moadqassem, done.

**Pull request merged**